CMake cleanup & MinGW support

crypto-ape opened the pull request:
Hey Monkeys!

I have cleaned up the CMake scripts for the main project also, and made some improvements:

- removed unused and irrelevant code
  - old and not supported Boost library handling routines, TCL and BDB references
- MinGW support ( supports building Windows binaries on Linux )
- enforcing more strict compiler warnings and errors

Python helpers:
 - rewritten `cat-parts` and `embed_genesis` helpers into Python
   - simplifies cross build
   - can be manually turned off with `GENERATING_HELPERS` CMake parameter
 - if Python is not found, the build continues in the usual way
 - works on Windows / Linux / Mac

I have also enabled code fortification mechanisms
 -  FORTIFY_SOURCE, safe-stack, stack-protector, ...

Related: https://github.com/bitshares/bitshares-fc/pull/121

Ape out!

pmconrad commented on CMakeLists.txt:
```diff
@@ -1,6 +1,6 @@
 # Defines BitShares library target.
-project( BitShares )
-cmake_minimum_required( VERSION 2.8.12 )
+cmake_minimum_required( VERSION 3.2 FATAL_ERROR )
```
Any specific reason for this? We usually avoid updating minimum requirements.

crypto-ape replied:
Yes, the same as in https://github.com/bitshares/bitshares-fc/pull/121

CMake that supports Visual Studio 2015 Generator.

pmconrad replied:
@abitmore @jmjatlanta what's the minimum cmake version we should support in your opinion / to your knowledge of windows/apple/linux build environments?

crypto-ape replied:
Another update,

The **first CMake** that is usable with the most up to date version of OpenSSL ( 1.0.2r ) is **3.1.2**.

Any version older than that will fail, with problems parsing version information from OpenSSL headers.

I have reworked the `FindPython` and `FindBoost` routines to be more future-proof.

Updates are coming soon.

pmconrad replied:
> The first CMake that is usable with the most up to date version of OpenSSL ( 1.0.2r ) is 3.1.2.

That's helpful, thanks!

pmconrad commented on CMakeLists.txt:
```diff
@@ -18,10 +18,87 @@ elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
+elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
+    # allow MSVC VS2015.1 and VS2013 toolchain ( VS2013 is not tested )
+    if ("${CMAKE_CXX_COMPILER_VERSION}" VERSION_EQUAL "19.0")
+        if (NOT CMAKE_CXX_COMPILER_VERSION VERSION_EQUAL "19.0.23506.0")
```
I'm not familiar with the windows world, but does it make sense to require such a specific version number?

crypto-ape replied:
This is the exact version of MS Visual C++ Compiler that comes with MS Visual Studio 2015 Update 1, which is to my knowledge the only version of MSVC that is able to compile the codebase.

jmjatlanta replied:
FYI: Support for the latest versions of MSVC are in code review.

crypto-ape replied:
That sounds great. Would it also support 2015, or 2017 only ? Or do you mean only the freshest 2019 ?

jmjatlanta replied:
I have not downloaded 2019 yet, but my guess is that it will work. I am still testing.

pmconrad commented on CMakeLists.txt:
```diff
@@ -18,10 +18,87 @@ elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
+    # allow MSVC VS2015.1 and VS2013 toolchain ( VS2013 is not tested )
+    if ("${CMAKE_CXX_COMPILER_VERSION}" VERSION_EQUAL "19.0")
+        if (NOT CMAKE_CXX_COMPILER_VERSION VERSION_EQUAL "19.0.23506.0")
+            message(FATAL_ERROR "Your MSVC version ${CMAKE_CXX_COMPILER_VERSION} does not match the required 19.0.23506.0 version ( Visual Studio 2015 Update 1 ).")
```
The message is wrong I think - according to the code we do support other versions, but not other 19.0.x versions.

crypto-ape replied:
Yes, you got it right :)

I did not want to be very restrictive, but I know from trial and error that for Visual Studio 2015 ( compiler version 19.0 ) the only working version is the specified one.
I have not tried older compilers, but I have a suspicion that it will not work due to missing C++11 features.

pmconrad commented on CMakeLists.txt:
```diff
@@ -50,37 +129,64 @@ LIST(APPEND BOOST_COMPONENTS thread
-   SET(Boost_LIBRARIES ${BOOST_LIBRARIES_TEMP} ${Boost_LIBRARIES})
-ENDIF()
+
+FIND_PACKAGE(Python 3.6 COMPONENTS Interpreter)
```
Would older versions work as well?

crypto-ape replied:
I have tested it only on Python 3.6+, but I guess even older Python 3 would work.
I would have to make sure, if you are interested in lowering the required version.

abitmore replied:
AFAICS Ubuntu 16.04 LTS ships with Python 2.7 and Python 3.5 by default. If 3 is required, I'd recommend that we check whether 3.5 would work.

pmconrad replied:
Just tested on an older system with `cmake-3.5` and both `python-2` and `python-3.4`. `/usr/bin/python` is `python-2` and `/usr/bin/python3` is `python-3`. I changed this line to require python 3.4, but the build fails with two errors:
```
CMake Error at CMakeModules/FindPython/Support.cmake:8 (cmake_policy):
  An attempt was made to set the policy version of CMake to "3.7" which is
  greater than this version of CMake.  This is not allowed because the
  greater version may have new policies not known to this CMake.  You may
  need a newer CMake version to build this project.
...
CMake Error at CMakeModules/FindPython/Support.cmake:226 (if):
  if given arguments:

    "_Python_VERSION" "VERSION_GREATER_EQUAL" "Python_FIND_VERSION"

  Unknown arguments specified
```

crypto-ape replied:
From my tests, the script is still usable with Python 3.3, and with a small two line modification, also with Python 3.2.

The CMake error comes from the FindPython CMake scripts taken from a recent CMake distribution.

I will try to make it usable with older CMake versions ( but for sure this won't support CMake 2 ), or the second option is, to skip Python checking in older CMake versions. In these cases, this would fall back to using the C++ build helpers, without the need for Python.

crypto-ape replied:
Another update,

The **first CMake** that is usable with the most up to date version of OpenSSL ( 1.0.2r ) **is 3.1.2**.

Any version older than that will fail, with problems parsing version information from OpenSSL headers.

I have reworked the `FindPython` and `FindBoost` routines to be more future-proof.

Updates are coming soon.

pmconrad commented on CMakeLists.txt:
```diff
@@ -50,37 +129,64 @@ LIST(APPEND BOOST_COMPONENTS thread
 
     if( MSVC )
-	add_definitions(-DWIN32_LEAN_AND_MEAN)
+	    add_definitions(-DWIN32_LEAN_AND_MEAN)
```
undo this whitespace change?

crypto-ape replied:
Sure, looks like a tab/space issue.

pmconrad commented:
What I really dislike about this PR is the huge load of new CMake modules. Is that really necessary?

I'm all for simplifying the build process, but adding 3kLoC only for replacing `cat-parts` and `embed_genesis` with python equivalents is not really it.

crypto-ape commented:
Well, I kind of agree with you. I was not sure about adding those files, neither.

In the meantime, I have come to an idea that would support older CMake 3 versions, and would not require so many supporting files.

This PR was not just about replacing `cat-parts`, and `embed_genesis`, but also about cleaning up the unneeded and confusing parts of CMake scripts.

abitmore commented:
Being able to "fallback" to old C++ `cat-parts` means we have to keep both the C++ code and the new python script in the code base? What will we gain by doing so?

crypto-ape commented:
> Being able to "fallback" to old C++ `cat-parts` means we have to keep both the C++ code and the new python script in the code base? What will we gain by doing so?

Yes, you are right. Both versions would have to be present.

Mostly, lowering requirements for the most common use case, making the build easier for the curious user.

The technical trad-off is of course the need to keep the both versions in sync.

On the other hand, it seems to me that this would be of no serious concern, because the functionality of `embed_genesis` and `cat-parts` is stable and would not require much modifications in the future.

crypto-ape commented:
Hey Monkeys, It's me again.

I was working the fix for the `fc` subroject, and now it is finally working.

Here is what is updated since the initial commit.

 - Updated `FindPython` and `FindBoost`, both are more future-proof
   - rewritten `FindPython` to be a lightweight wrapper for `FindPythonInterp`, if `FindPython` is not available ( CMake < 3.12 )
 - FindBoost
   - uses config from `fc` subproject
     - the configuration is in one place
   - the `fc` does some FindBoost magic
 - parallel (multiprocessed) `embed_genesis` Python script - speed improvement while processing large genesis files

This project requires updated version of `fc` to compile (namely the `BoostConfig.cmake` file)

abitmore commented:
Travis complained:
```
CMake Error at CMakeLists.txt:116 (add_subdirectory):
  add_subdirectory given source "libraries/fc/CMakeModules/Legacy" which is
  not an existing directory.
CMake Error at CMakeLists.txt:147 (FIND_PACKAGE):
  Could not find a package configuration file provided by "Boost" with any of
  the following names:
    BoostConfig.cmake
    boost-config.cmake
  Add the installation prefix of "Boost" to CMAKE_PREFIX_PATH or set
  "Boost_DIR" to a directory containing one of the above files.  If "Boost"
  provides a separate development package or SDK, be sure it has been
  installed.
```

pmconrad commented:
See comment above:
> This project requires updated version of fc to compile (namely the BoostConfig.cmake file)

pmconrad commented:
https://github.com/bitshares/bitshares-fc/pull/121 has been merged, please rebase and bump fc.

crypto-ape commented:
I have failed to compile the most up-to-date, `develop` with `fc:master`, branch, because of issues unrelated to this PR.


There is some other refactoring in the `fc` submodule which probably causes this failure.


Excerpt from compilation log:
```
libraries/protocol/include/graphene/protocol/object_id.hpp:171:17: error: ‘true_type’ in namespace ‘fc’ does not name a type
     typedef fc::true_type  is_defined;
                 ^~~~~~~~~
/libraries/protocol/include/graphene/protocol/object_id.hpp:172:17: error: ‘false_type’ in namespace ‘fc’ does not name a type
     typedef fc::false_type is_enum;
```

After fixing this, another issues with `uint128` arose.

It seems to me, that the `develop` branch in its current form is not usable with the `fc:master` branch, and consequently neither is this PR branch.

pmconrad commented:
Right, sorry. `fc/master` requires changes from #1789 . There are still some issues in there wrt windows + mac builds, that's why it hasn't been merged into develop yet.

pmconrad commented:
@crypto-ape please rebase on latest develop. #1789 has been merged in so there shouldn't be conflicts anymore.

crypto-ape commented:
@pmconrad Working on it. Will push tomorrow If nothing unexpected happens.

oxarbitrage commented:
@crypto-ape any chance to get this done today ? we will want to include this in the current release that we are closing today. I can try to fix conflicts but will be better if you can take it. 

Thank you very much and let us know.

crypto-ape commented:
@oxarbitrage I have pushed the changes, hopefully it will build.
Later, for sure I will have to commit a fix for Windows & MinGW.

Otherwise, 
Linux ,and Windows with MSVC should work well.

crypto-ape commented:
@oxarbitrage, @pmconrad It is done, hopefully it is enough for now.

oxarbitrage commented:
@crypto-ape , my apologies. We decided to move it to the next one as there are some Windows specifics problems at building and we are still unsure about the introduction of new python to the core. 

Those deserves some more time for discussion and review and that's the reason of the move.

crypto-ape commented:
@oxarbitrage Not a problem. I have the same opinion, it should not be rushed.

I would not mind, If you would consider me taking part in the discussion process.

pmconrad commented:
@crypto-ape please check #1923 - it uses CMake-builtins for the generated files instead of requiring python. What do you think?

crypto-ape commented:
@pmconrad https://github.com/bitshares/bitshares-core/pull/1923 looks good to me, it does the job.

As you mention, it has the one drawback that you have to manually re-run CMake if needed.

The ideal solution of course would be, if the CMake file generation could be invoked automatically after the relevant files are changed. Then the C++ build tools could be removed.

abitmore commented:
>The ideal solution of course would be, if the CMake file generation could be invoked automatically after the relevant files are changed. 

I guess CMake can check that.

pmconrad commented:
Hi @crypto-ape , we have merged #1923 with the cmake-based build tools. Please remove the python stuff and rebase on latest develop. Thanks!

crypto-ape commented:
Hello friends,

I have removed the Python scripting completely and rebased the branch on top of develop.

The Linux tool-chain works, but:
 
  - there is a problem with the `fc` library in `tests/reflection_tests.cpp` when compiling with MinGW ( and maybe MSVC )
    - ambiguous `dispatch` call; can be 'fixed' easily
	```cpp
	BOOST_CHECK_EQUAL(fc::typelist::runtime::dispatch(list(), (std::size_t)0ul, get_name), "float"); // force dispatch with std::size_t
	```
  - MinGW fails to compile cli_wallet, because `wallet.cpp` is somehow bloated and the resulting object file `wallet.obj` exceeds limit on the length of string table
	
  - MSVC works
  
The problems are out of the scope of this PR so I would suggest merging this to develop and resolving the issues in separate PRs.

After I am done with testing, I will push. ETA is sometime today.

abitmore commented:
Issue https://github.com/bitshares/bitshares-core/issues/950 and PR https://github.com/bitshares/bitshares-core/pull/1413 were for splitting `wallet.cpp` into multiple files, but staled.

pmconrad commented:
> MinGW fails to compile cli_wallet

You mean MinGW on linux? Which version? (Haven't had problems with that in bitshares-gitian.)

crypto-ape commented:
> > MinGW fails to compile cli_wallet
> 
> You mean MinGW on linux? Which version? (Haven't had problems with that in bitshares-gitian.)

It fails on both Linux and Windows. It is a problem of the `develop` branch. 
Before, it compiled. The current `develop` branch fails to build with MinGW.

I build with MinGW 7.3 both on Linux Ubuntu 18.04 and Windows 10.

abitmore commented:
It's likely that we'll process this PR after merged #1992 which bumps FC, so the FC bump in this PR will be unneeded.

pmconrad commented:
Merged #1992, please rebase on latest develop (again, sorry) and get rid of the fc bump.
Will test develop build with gitian.

pmconrad commented on libraries/CMakeLists.txt:
```diff
@@ -1,3 +1,16 @@
+if( MINGW )
```
I find it confusing to have CURL_LIBRARIES and CURL_ADDITIONAL_LIBS, and to have them detected / modified / used differently in different places. Perhaps unify and move this into top-level CMakeLists.txt?

pmconrad commented on CMakeModules/Utils.cmake:
```diff
@@ -11,3 +11,36 @@ macro(add_linker_flag _FLAG)
+endmacro(add_compiler_flag_if_available _VAR_NAME _FLAG)
+
+# macro to verify if required flags are set. If not then the project should be skipped
+macro(verify_required_flags _FLAGS)
```
Unused, please remove.

pmconrad commented on CMakeModules/Utils.cmake:
```diff
@@ -11,3 +11,36 @@ macro(add_linker_flag _FLAG)
+	endif()
+endmacro()
+
+macro(link_libraries_by_flag _FLAG _LIST _LIB)
```
Unused, please remove.

pmconrad commented on CMakeLists.txt:
```diff
@@ -258,6 +270,10 @@ else( WIN32 ) # Apple AND Linux
 endif( WIN32 )
 
+if ( FULL_STATIC_BUILD )
+    set( CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -static -static-libstdc++ -static-libgcc" )
```
Won't this break if FULL_STATIC_BUILD is set with MSVC for example?

crypto-ape replied:
Will add condition, to make sure it won't get triggered for MSVC.

pmconrad commented on CMakeLists.txt:
```diff
@@ -28,6 +28,16 @@ elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
         message(FATAL_ERROR "Clang version must be at least 3.3!")
     endif()
+elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
+    # allow MSVC VS2015.1 and VS2013 toolchain ( VS2013 is not tested )
```
ISTR this PR predates our switch to C++-14. I suppose VS2013 does not support C++-14, so perhaps this check can be simplified?

crypto-ape commented:
@pmconrad please review, updated according your suggestions.

I have tested, seems to work on all platforms, except the problem with `cli_wallet` where the `wallet.cpp` fails for MinGW.

pmconrad commented on libraries/utilities/CMakeLists.txt:
```diff
@@ -30,7 +30,7 @@ if(CURL_STATICLIB)
   COMPILE_DEFINITIONS "CURL_STATICLIB")
 endif(CURL_STATICLIB)
-target_link_libraries( graphene_utilities fc ${CURL_LIBRARIES})
+target_link_libraries( graphene_utilities fc ${CURL_LIBRARIES} ${CURL_ADDITIONAL_LIBS} )
```
`CURL_ADDITIONAL_LIBS` is obsolete now

crypto-ape replied:
Code updated

pmconrad commented:
Thanks!
